controllers/controlplane_controller_utils.go

```diff
@@ -42,12 +42,63 @@ func setControlPlaneDefaults(spec *operatorv1alpha1.ControlPlaneDeploymentOption
 
 	if spec.DataPlane != nil && *spec.DataPlane != "" {
 		if _, isOverrideDisabled := dontOverride["CONTROLLER_PUBLISH_SERVICE"]; !isOverrideDisabled {
-			spec.Env = updateEnv(spec.Env, "CONTROLLER_PUBLISH_SERVICE", fmt.Sprintf("%s/svc-%s", namespace, *spec.DataPlane))
+			spec.Env = updateEnv(spec.Env, "CONTROLLER_PUBLISH_SERVICE", controllerPublishService(*spec.DataPlane, namespace))
 		}
 		if _, isOverrideDisabled := dontOverride["CONTROLLER_KONG_ADMIN_URL"]; !isOverrideDisabled {
-			spec.Env = updateEnv(spec.Env, "CONTROLLER_KONG_ADMIN_URL", fmt.Sprintf("https://svc-%s.%s.svc:%d", *spec.DataPlane, namespace, dataplaneutils.DefaultKongAdminPort))
+			spec.Env = updateEnv(spec.Env, "CONTROLLER_KONG_ADMIN_URL", controllerKongAdminURL(*spec.DataPlane, namespace))
 		}
 	}
+
+}
+
+func setControlPlaneEnvOnDataPlaneChange(
+	spec *operatorv1alpha1.ControlPlaneDeploymentOptions,
+	namespace string,
+) bool {
+	var changed bool
+
+	dataplaneIsSet := spec.DataPlane != nil && *spec.DataPlane != ""
+	if dataplaneIsSet {
+		newPublishServiceValue := controllerPublishService(*spec.DataPlane, namespace)
+		if envValue(spec.Env, "CONTROLLER_PUBLISH_SERVICE") != newPublishServiceValue {
+			spec.Env = updateEnv(spec.Env, "CONTROLLER_PUBLISH_SERVICE", newPublishServiceValue)
+			changed = true
+		}
+		newKongAdminURL := controllerKongAdminURL(*spec.DataPlane, namespace)
+		if envValue(spec.Env, "CONTROLLER_KONG_ADMIN_URL") != newKongAdminURL {
+			spec.Env = updateEnv(spec.Env, "CONTROLLER_KONG_ADMIN_URL", newKongAdminURL)
+			changed = true
+		}
+	} else {
+		if envValue(spec.Env, "CONTROLLER_PUBLISH_SERVICE") != "" {
+			spec.Env = rejectEnv(spec.Env, "CONTROLLER_PUBLISH_SERVICE")
+			changed = true
+		}
+		if envValue(spec.Env, "CONTROLLER_KONG_ADMIN_URL") != "" {
+			spec.Env = rejectEnv(spec.Env, "CONTROLLER_KONG_ADMIN_URL")
+			changed = true
+		}
+	}
+
+	return changed
+}
+
+func controllerKongAdminURL(dataplaneName, dataplaneNamespace string) string {
+	return fmt.Sprintf("https://svc-%s.%s.svc:%d",
+		dataplaneName, dataplaneNamespace, dataplaneutils.DefaultKongAdminPort)
+}
+
+func controllerPublishService(dataplaneName, dataplaneNamespace string) string {
+	return fmt.Sprintf("%s/svc-%s", dataplaneNamespace, dataplaneName)
+}
+
+func envValue(env []corev1.EnvVar, name string) string {
```

Comment:
```suggestion
func envValueByName(env []corev1.EnvVar, name string) string {
```

Perhaps this could help to make it obvious what this `func` does?